### PR TITLE
Mount Data and Metadata on provided partition with systemd + Best practices

### PR DESCRIPTION
**Pull request by Gredin67**
## Problem
See https://garagehq.deuxfleurs.fr/documentation/cookbook/real-world/#best-practices

- Metadata should be stored on a partition with checksumming and integrity verification, especially if using LMDB
- If a huge quantity of data is to be stored, then Data should be stored on XFS filesystem
- Metadata take about 1% of the Data size. More if lots of small data is stored
- Use LMDB database that is more tested

## Solution

- Add doc to help user create appropriate partitions and size
- ask for partition to be used for data on a SSD or HDD, `/dev/sdb1`
- ask for partition to be used for metadata, e.g. `/dev/sda2`

## PR Status

- [x] Use LMDB database rather than SQLite
- [x] keep metadata on YunoHost partition `/dev/sda1` if it is already ZFS or BTRFS on a SSD
- [x] Set `block_size = "10M"` if all nodes have FTTH connectivity
- [x] Decide if asking for `/dev/sdb1` or UUID in manifest
- [x] Add checksumming and integrity verification of the Metadata partition.  This can be done using Garage's built-in automatic snapshotting (since v0.9.4), or by using filesystem level snapshots --> https://garagehq.deuxfleurs.fr/documentation/reference-manual/configuration/#metadata_auto_snapshot_interval
- [x] merge and align with https://github.com/YunoHost/yunohost/pull/2230
- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by Gredin67** [comment]
!testme

**Comment by Tagadda** [comment]
!testme

**Comment by Gredin67** [comment]
@Tagadda  this PR needs testing with bare metal hardware to test mounting of partitions with systemd 
I still encounter what seems to be permission issues after mount, see https://paste.yunohost.org/raw/urapemumih

**Comment by Gredin67** [comment]
!testme

**Comment by Gredin67** [comment]
wait for https://github.com/YunoHost/yunohost/pull/2282 to be released

**Comment by Gredin67** [comment]
!testme

File: doc/ADMIN.md [hunk 1]
```diff
@@ -1,27 +1,119 @@
...
 
 # Informations you should be aware of:
 
- * This application provide a node that you can connect to a garage cluster. Few option are manageable by the config panel for the current node but IT DOESN'T offer simpler way to manage bucket and keys. You have to do it by command line or let an other node managing it.
- * This application consider that the weight of the node is the size reserved to garage in G (Gigabytes)
- * This application will try to create a virtual disk to ensure garage doesn't use more than allowed. If virtualisation is not available your responsible to check space used by garage.
- * To connect from another node, you might need the RPC port. Is is defined in `rpc_bind_addr` in your `garage.toml`
- * Storage consumption : appart from your data storage, you can expect the metadata (database) to consume approximatly 1% of the data size (1GB for 100GB of data for instance), or a bigger value if you store many small objects.
+* This application provides a Node that you can connect to a Garage Cluster. The Config Panel provides some Node-related settings. Cluster-related settings like Key or Bucket should be managed via Garage CLI or by another Node
```
Comment (halfa) [inline doc/ADMIN.md:10]: We have a UI coming soon that could allow you do those operations, and there are at least two community ones
- https://github.com/Noooste/garage-ui
- https://github.com/khairul169/garage-webui

File: doc/ADMIN.md [hunk 2]
```diff
@@ -1,27 +1,119 @@
...
+* Garage does not set a hard storage Quota. So disk space usage can become greater than the Node Weight you define. 
+* To prevent this, this App will ask you for a Data partition at install, or create a virtual disk with size the Weight of your Node. Virtual disk is only created if YunoHost is not in a VM, and performance may be poor.
+* If you install on a VM and do not provide a Data partition, you are responsible for checking sufficient available storage.
+* Expect Metadata (database) and its 2-3 snapshots to be approximatly 3% of the Data size (3GB for 100GB of Data for instance), or a bigger percentage if you store many small objects.
```
Comment (halfa) [inline doc/ADMIN.md:14]: Objects below 3kb will be stored directly in the metadata database

File: doc/ADMIN.md [hunk 3]
```diff
@@ -1,27 +1,119 @@
...
+* To connect from another Node, you might need the RPC port. Is is defined in `rpc_bind_addr` in your `garage.toml`
```
Comment (halfa) [inline doc/ADMIN.md]:
```suggestion
* To connect from another Node, you will need the RPC port. Is is defined in `rpc_bind_addr` in your `garage.toml`
```

File: doc/ADMIN.md [hunk 4]
```diff
@@ -1,27 +1,119 @@
...
+  * Recovery from other Nodes within a day (no Data snapshot)
+* Data-center-like
+  * Unclean shutdown unlikely
+  * Fast and high-service networking
+  * Data storage greater than some TB
+  * Usage for high-performance distributed storage
+  * Local Recovery of Node within minutes, e.g. with Data snapshot
+
+**Recommended** (minimal) self-hosting config:
+* Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too (EXT4 only for storage in the GB order)
```
Comment (halfa) [inline doc/ADMIN.md:53]:
```suggestion
* Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too (EXT4 can limit you due to the restricted amount of inodes available)
```

File: doc/ADMIN.md [hunk 5]
```diff
@@ -1,27 +1,119 @@
...
+* Metadata partition: on **SSD** (HDD OK if lots of RAM for kernel caching), **BTRFS or ZFS** (EXT4 OK with Garage-Snapshot)
+* Database: **LMDB is default, more tested, more performant. Recommended if Metadata on HDD**. (Prefer SQLite on a 32-bit system if you plan to store much data.)
```
Comment (halfa) [inline doc/ADMIN.md]:
```suggestion
* Database: **LMDB is default, more tested, more performant assuming stable power and reliable storage**. Otherwise use SQLite. Always use SQLite on a 32 bit system.
```

File: doc/ADMIN.md [hunk 6]
```diff
@@ -1,27 +1,119 @@
...
+* Metadata may get corrupt, e.g. after a power outage: **Setup Garage-Snapshot** (SQLite should be avoided, but can be more robust with frequent unclean shutdown)
```
Comment (halfa) [inline doc/ADMIN.md]:
```suggestion
* Metadata may get corrupt, e.g. after a power outage: **Setup Garage-Snapshot** (SQLite is more resilient to frequent unclean shutdown)
```

File: doc/ADMIN.md [hunk 7]
```diff
@@ -1,27 +1,119 @@
...
+* (Use SQLite if you want to be able to migrate Metadata to a different architecture without resyncing, e.g. from AMD64 to ARM64.)
+* `blocksize = "10M"` if you have FTTH and plan to store mostly large files, **leave blocksize to default otherwise**
```
Comment (halfa) [inline doc/ADMIN.md]: There are relatively few downsides to increasing the blocks size. Block size is actually a "max" block size, any file _above_ this limit will be cut in multiple files. For example if the block size is set to 1MB (default), then a 1.5MB file will be stored in two blocks of 1MB and 0.5MB. But if you set the block_size to 10MB, it will result in a single 1.5MB block. In practice, the blocks are compressed, so they will rarely reach the actual block size on disk.

The downsides of increasing the blocks size are increased memory usage and longer transfer time between nodes (RPC operations). They are relatively minor, so unless you operate in very limited memory and networking, raising the block size can benefit you greatly or even save you significant overhead, especially if your content is big (video or heavy images).

File: doc/ADMIN.md [hunk 8]
```diff
@@ -1,27 +1,119 @@
...
+
+**Recommended** (minimal) data-center config:
+* Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too
+* Metadata partition: on **SSD**. **BTRFS or ZFS with Garage-Snapshot** (`ext4` with Garage-Snapshot)
+* (You may replace Garage-Snapshot with filesystem snapshots if you stop garage during the snapshot)
+* Database: **LMDB** (SQLite if on 32-bit architecture)
```
Comment (halfa) [inline doc/ADMIN.md]: No datacenter-like config should use 32 bit systems nowdays 😅

File: doc/ADMIN.md [hunk 9]
```diff
@@ -1,27 +1,119 @@
...
+
+## Simple self-hosting Backups use-case
+
+* Yunohost installed on a 60GB SSD (NVME or SATA)
+* `/` is mounted on `ext4` partition `/dev/sda1` and so is `/home/yunohost.app/garage/`
+* An empty 16TB HDD `/dev/sdb` (USB3 or SATA, USB2 should be avoided)
+* Create a 16TB partition `/dev/sdb1` for the Data
+* Type `/dev/sdb1` for the Data and `no` for the Metadata at install
+* `/dev/sdb1` will be formatted in `xfs` and mounted on `/home/yunohost.app/garage/data`
+* Garage will take Snapshots of your LMDB database regularly
```
Comment (halfa) [inline doc/ADMIN.md]: You might want to replace this with SQLite, especially if you have a single node

File: doc/ADMIN.md (hunk 1 above)
Comment (Gredin67) [reply doc/ADMIN.md:10]: OK then we will not spend much time working on the garage_ynh config panel. Will it be integrated in garage or will need to be installed as a separate package ?

File: doc/ADMIN.md (hunk 4 above)
Comment (Gredin67) [reply doc/ADMIN.md:53]: Suppose yunohost users don't now what inodes are. What is the inode limit ? 
I prefer to give hints that can help users decide depending on their use-case.  Like a rough number of files and number of GB above which not to use ext4. 1, 10, 100, 1000 GB ? 1 10TB ?

File: doc/ADMIN.md (hunk 7 above)
Comment (Gredin67) [reply doc/ADMIN.md]: what about changing the default value to 10MB upstream then ?

File: doc/ADMIN.md (hunk 8 above)
Comment (Gredin67) [reply doc/ADMIN.md]: yes, I'll remove this

File: doc/ADMIN.md (hunk 9 above)
Comment (Gredin67) [reply doc/ADMIN.md]: is ext4 OK in this use-case or should we say metadata on ext4 is only OK for e.g. 4TB of data max ?

File: doc/ADMIN.md (hunk 1 above)
Comment (Gredin67) [reply doc/ADMIN.md:10]: https://github.com/YunoHost-Apps/garage_ynh/issues/62

File: doc/ADMIN.md (hunk 7 above)
Comment (Gredin67) [reply doc/ADMIN.md]: I've set `10MB` default everywhere

File: doc/ADMIN.md (hunk 9 above)
Comment (Gredin67) [reply doc/ADMIN.md]: I've set SQLite SQLite

File: scripts/install [hunk 10]
```diff
@@ -7,25 +7,88 @@ source /usr/share/yunohost/helpers
...
 else
 	rpc_secret=$(ynh_string_random --length=32| od -A n -t x -w64 | sed 's/ //g')
 fi
 if [ -n "$bootstrap_peers" ]
-
 then
 	echo "$bootstrap_peers" | grep -E '[0-9a-f]{64}@((\b25[0-5]|\b2[0-4][0-9]|\b[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}|([a-zA-Z0-9][a-zA-Z0-9-]{0,61}[a-zA-Z0-9]\.)+[a-zA-Z]{2,}):[0-9]{1,4}' || ynh_die "friend server id must have id with the following form : 1799bccfd7411eddcf9ebd316bc1f5287ad12a68094e1c6ac6abde7e6feae1ec@192.168.1.1:1234 or  1799bccfd7411eddcf9ebd316bc1f5287ad12a68094e1c6ac6abde7e6feae1ec@example.tld:1234"
 fi
 
+#=================================================
```
Comment (Gredin67) [inline scripts/install:34]: @halfa what do you think of those performance tiers ? Would you think of other settings we could adjust here ?
Maybe the filesystem type ? or the meta-data snapshot strategy ?
Also testing for maximum weight depending on DB and FS type ?

File: doc/ADMIN.md (hunk 1 above)
Comment (halfa) [reply doc/ADMIN.md:10]: The two community ones need to be setup as a separate program, the UI that we plan will be embedded into garage

File: doc/ADMIN.md (hunk 4 above)
Comment (halfa) [reply doc/ADMIN.md:53]: It's hard to predict because the number of inodes on an ext4 volume will depend on the volume size (see [this answer](https://unix.stackexchange.com/questions/785941/has-the-default-value-for-bytes-per-inode-changed)), and garage's filecount will depends on file size, configured block_size setting and object count.

For example uploading a 1GB video file will result (with the default 1MB block_size) in 1024 files on disk.

However we can make some napkin calculations: assuming ext4 defaults, for a volume below 4TB, we have 16,384 bytes per inode. So as long as file are bigger than 16k in their compressed state, you'll fill up the volume before running out of inodes.

File: doc/ADMIN.md (hunk 7 above)
Comment (halfa) [reply doc/ADMIN.md]: > what about changing the default value to 10MB upstream then ?

It's something we can consider.

File: scripts/install (hunk 10 above)
Comment (halfa) [reply scripts/install:34]: Not sure if those brackets are a yunohost typical use case, but I would suggest:

- data-center/high performance self-hosted - reliable power (UPS) and network: lmdb + 50MB
- self-host single node - good network, no UPS (x64): sqlite + 10MB
- embedded (ARM, 32bit...) or unreliable power:  sqlite + 1MB
